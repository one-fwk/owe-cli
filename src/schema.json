{
  "additionalProperties": false,
  "required": ["projects", "browserTargets"],
  "properties": {
    "browserTargets": {
      "type": "array"
    },
    "projects": {
      "type": "object"
    }
  }
}