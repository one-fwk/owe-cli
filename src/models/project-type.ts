export enum ProjectType {
  LIBRARY = 'library',
  EXTENSION = 'extension',
}