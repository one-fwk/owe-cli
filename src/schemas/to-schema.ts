import { join } from 'path';

type Const<T> = { const: T };
type Enum<T> = { enum: T };

export function toSchemaConstants<T>(obj: Object): Const<T>[] {
  return (<any>Object).values(obj).map((value: T) => ({
    const: value,
  }));
}

export function toSchemaEnum<T>(obj: Object): Enum<T>[] {
  return [
    {
      enum: (<any>Object).values(obj),
    },
  ];
}

export class MockSchema {
  constructor(
    private readonly schema: string,
    private readonly mockSchemas: string[],
  ) {}

  private getPath(schema: string) {
    const paths = schema.split('#');

    return {
      path: join(__dirname, paths[0]),
      exported: paths[1],
    };
  }

  public async mock<T>(): Promise<T> {
    this.mockSchemas.forEach(schema => {
      const { path, exported } = this.getPath(schema);

      jest.doMock(path, () => ({
        [exported]: {
          type: 'object',
        },
      }));
    });

    const { path, exported } = this.getPath(this.schema);
    return (await import(path))[exported];
  }

  public unmock() {
    this.mockSchemas.forEach(schema => {
      jest.unmock(this.getPath(schema).path);
    });
  }
}