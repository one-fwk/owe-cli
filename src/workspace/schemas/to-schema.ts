import { join } from 'path';

export type Const<T> = { const: T };
export type Enum<T> = { enum: T };
export type Type = 'string' | 'object' | 'array' | 'number' | 'boolean';

export const toType = (type: Type) => (schema: any = {}) => ({ ...schema, type });

export const toTypeString = toType('string');
export const toTypeArray = toType('array');
export const toTypeObject = toType('object');
export const toTypeNumber = toType('number');

export const uniqueArray = (items: any) => (schema = {}) => toTypeArray({
  uniqueItems: true,
  items,
  ...schema,
});

export const filePathSchema = (errorMessage: string) => toTypeString({
  regexp: '/^[a-z]:((\\/|(\\\\?))[\\w .]+)/i',
  errorMessage,
});

export const jsonFileSchema = (errorMessage: string) => toTypeString({
  regexp: '/^[a-z]:((\\/|(\\\\?))[\\w .]+)+\\.json/i;',
  errorMessage,
});

export const uniqueArrayOfStrings = uniqueArray(toTypeString());

export function toSchemaConstants<T>(obj: Object): Const<T>[] {
  return (<any>Object).values(obj).map((value: T) => ({
    const: value,
  }));
}

export function toSchemaEnum<T>(obj: Object): Enum<T>[] {
  return [
    {
      enum: (<any>Object).values(obj),
    },
  ];
}

export class MockSchema {
  constructor(
    private readonly schema: string,
    private readonly mockSchemas: string[],
  ) {}

  private getPath(schema: string) {
    const paths = schema.split('#');

    return {
      path: join(__dirname, paths[0]),
      exported: paths[1],
    };
  }

  public async mock<T>(): Promise<T> {
    this.mockSchemas.forEach(schema => {
      const { path, exported } = this.getPath(schema);

      jest.doMock(path, () => ({
        [exported]: {
          type: 'object',
        },
      }));
    });

    const { path, exported } = this.getPath(this.schema);
    return (await import(path))[exported];
  }

  public unmock() {
    this.mockSchemas.forEach(schema => {
      jest.unmock(this.getPath(schema).path);
    });
  }
}