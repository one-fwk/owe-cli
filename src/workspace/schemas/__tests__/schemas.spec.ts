import { ValidateFunction } from 'ajv';
import * as Ajv from 'ajv';

import { filePathSchema, MockSchema } from '../to-schema';
import { projectSchema } from '../project.schema';
import { projectTypeSchema } from '../project-type.schema';
import { browserTargetsSchema } from '../browser-targets.schema';

describe('Schema validation', () => {
  let ajv: Ajv.Ajv;

  beforeEach(() => {
    ajv = new Ajv({
      useDefaults: true,
    });
  });

  describe('projects', () => {
    let workspaceSchema: any;
    let validate: ValidateFunction;
    let schema: MockSchema;

    const createTestProject = (name: string) => ({
      projects: {
        [name]: {},
      },
    });

    beforeEach(async () => {
      schema = new MockSchema(
        'owe.schema#workspaceSchema',
        ['project.schema#projectSchema']
      );

      workspaceSchema = await schema.mock();
    });

    afterAll(() => {
      schema.unmock();
    });

    it('should accept name with underscore', () => {
      expect(ajv.validate(
        workspaceSchema,
        createTestProject('test_app')),
      ).toBe(true);
    });

    it('should accept name with dash', () => {
      expect(ajv.validate(
        workspaceSchema,
        createTestProject('test-app')),
      ).toBe(true);
    });

    it(`shouldn't accept illegal characters in name`, () => {
      expect(ajv.validate(
        workspaceSchema,
        createTestProject('test app # %')),
      ).toBe(false);

      expect(ajv.errors).toMatchSnapshot();
    });
  });

  describe('schemas', () => {
    it('project schema should accept properties', () => {
      expect(ajv.validate(projectSchema,{
        sourceRoot: 'dist/appwriter',
        browserTargets: ['chrome'],
        projectType: 'extension',
      })).toBe(true);
    });

    describe('sourceRoot', () => {
      it('should fail if not valid path', () => {
        // doesn't work
        require('ajv-keywords')(ajv, ['regexp']);
        expect(ajv.validate(
          filePathSchema(''),
          'dist/appwriter?cloud'),
        ).toBe(false);

        expect(ajv.errors).toMatchSnapshot();
      });
    });

    describe('projectType', () => {
      it('should accept one of ProjectType values', () => {
        ['library', 'extension'].forEach(type => {
          expect(ajv.validate(projectTypeSchema, type)).toBe(true);
        });
      });

      it('should fail if empty', () => {
        expect(ajv.validate(projectTypeSchema,'')).toBe(false);
        expect(ajv.errors).toMatchSnapshot();
      });
    });

    describe('browserTargets', () => {
      it('should accept BrowserTarget values', () => {
        [['chrome'], ['chrome', 'edge']].forEach(targets => {
          expect(ajv.validate(browserTargetsSchema, targets)).toBe(true);
        });
      });

      it('should fail if value is not in BrowserTarget', () => {
        expect(ajv.validate(browserTargetsSchema, ['safari'])).toBe(false);
        expect(ajv.errors).toMatchSnapshot();
      });

      it('should fail if empty', () => {
        expect(ajv.validate(browserTargetsSchema, [])).toBe(false);
        expect(ajv.errors).toMatchSnapshot();
      });

      it('should fail if duplicate targets', () => {
        expect(ajv.validate(browserTargetsSchema, ['chrome', 'chrome'])).toBe(false);
        expect(ajv.errors).toMatchSnapshot();
      });
    });
  });
});