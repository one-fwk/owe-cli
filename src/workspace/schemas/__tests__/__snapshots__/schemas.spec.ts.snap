// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema validation schemas sourceRoot should fail if not valid path 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "regexp",
    "message": "should pass \\"regexp\\" keyword validation",
    "params": Object {
      "keyword": "regexp",
    },
    "schemaPath": "#/regexp",
  },
]
`;
